presto-hive/src/main/java/com/facebook/presto/hive/SubfieldExtractor.java

```diff
@@ -81,6 +87,8 @@ public SubfieldExtractor(StandardFunctionResolution functionResolution)
             if (expression instanceof CallExpression && functionResolution.isSubscriptFunction(((CallExpression) expression).getFunctionHandle())) {
                 List<RowExpression> arguments = ((CallExpression) expression).getArguments();
                 RowExpression indexExpression = arguments.get(1);
+                indexExpression = expressionOptimizer.apply(indexExpression);
```

Comment:
combine this statement with the one above: `RowExpression indexExpression = expressionOptimizer.apply(arguments.get(1));
`